### PR TITLE
Add optional `allow_tensor_metadata_change` flag to detach()

### PR DESCRIPTION
**Pull request by yf225**
When working on https://github.com/pytorch/pytorch/pull/22172, we found that there is a need to add an optional `allow_tensor_metadata_change` flag to `tensor.detach()`, for use cases where we do want to be able to change the detached tensor's sizes/strides after `detach()`. This PR adds the flag.

### CODE REVIEW

**Review by gchanan — COMMENTED** [review]
is there a reason to do things this way over the `__future__` solution as with get_overwrite_module_params_on_conversion?

**Comment by yf225** [comment]
@gchanan I think in the most general case, preventing changes to detached tensors' metadata is still useful because it's more restrictive. We need to add `allow_tensor_metadata_change` flag to `detach()` only because for some of the call sites of `detach()` we need a convenient way to also specify that `allow_tensor_metadata_change` should be true, so deciding whether to set this flag should be a local decision, and shouldn't affect how other `detach()` calls in the code behave (which would be the case if we use a global flag in `__future__`).

**Comment by gchanan** [comment]
But what's the underlying reason for preventing metadata changes?
